// nuxt-tradingview/dist/module.mjs
import { defineNuxtModule, createResolver, addComponent } from '@nuxt/kit';

const module = defineNuxtModule({
  meta: {
    name: "nuxt-tradingview",
    configKey: "tradingview",
    compatibility: {
      nuxt: ">=3.0.0"
    }
  },
  defaults: {
    prefix: "",
    importOnly: void 0,
    experimental: {
      anonymousCrossOrigin: false
    }
  },
  setup(options, nuxt) {
    const { resolve } = createResolver(import.meta.url);
    const components = [
      "SingleTicker",
      "Ticker",
      "SymbolInfo",
      "StockMarket",
      "StockHeatMap",
      "Screener",
      "MiniChart",
      "MarketOverview",
      "MarketData",
      "TickerTape",
      "FundamentalData",
      "ForexHeatMap",
      "ForexCrossRates",
      "EconomicCalendar",
      "TechnicalAnalysis",
      "CryptoMarket",
      "CryptoHeatMap",
      "CompanyProfile",
      "Chart",
      "SymbolOverview",
      "TopStories"
    ];
    const processComponent = (component) => {
      return {
        name: `${options.prefix}${component}`,
        global: true,
        filePath: resolve(`./runtime/components/${component}.vue`)
      };
    };
    const importWidgets = options.importOnly ? components.filter(
      (component) => options.importOnly?.includes(component)
    ) : components;
    importWidgets.map(processComponent).forEach(addComponent);
    nuxt.options.runtimeConfig.public.tradingview = {
      ...nuxt.options.runtimeConfig.public.tradingview,
      experimental: {
        anonymousCrossOrigin: !!options.experimental?.anonymousCrossOrigin
      }
    };
  }
});

export { module as default };
